File: app/src/main/java/droiddevelopers254/droidconke/views/activities/SessionFeedBackActivity.java
package droiddevelopers254.droidconke.views.activities;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import droiddevelopers254.droidconke.R;
import droiddevelopers254.droidconke.models.SessionsModel;
import droiddevelopers254.droidconke.models.SessionsUserFeedback;
import droiddevelopers254.droidconke.viewmodels.SessionDataViewModel;

public class SessionFeedBackActivity extends AppCompatActivity {

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.txtSessionFeedbackTitle)
    TextView txtSessionFeedbackTitle;
    @BindView(R.id.txtSessionUserFeedback)
    TextInputEditText txtSessionUserFeedback;
    @BindView(R.id.fab)
    FloatingActionButton fab;

    SessionDataViewModel sessionDataViewModel;
    SessionsModel sessionsModel1;
    @BindView(R.id.loginProgressBarFeedBack)
    ProgressBar loginProgressBarFeedBack;
    private int sessionId;
    private String dayNumber,sessionFeedback;
    SessionsUserFeedback userFeedback;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_session_feed_back);
        ButterKnife.bind(this);
        setSupportActionBar(toolbar);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);

            getSupportActionBar().setTitle("Sessions Feedback");
        }

        //get extras
        Intent extraIntent = getIntent();
        sessionId = extraIntent.getIntExtra("sessionId", 0);
        dayNumber = extraIntent.getStringExtra("dayNumber");

        sessionDataViewModel = ViewModelProviders.of(this).get(SessionDataViewModel.class);

        getSessionData(dayNumber, sessionId);

        //observe live data emitted by view model
        sessionDataViewModel.getSessionData().observe(this, sessionDataState -> {
            assert sessionDataState != null;
            if (sessionDataState.getDatabaseError() != null) {
                handleDatabaseError(sessionDataState.getDatabaseError());
            } else {
                this.handleFetchSessionData(sessionDataState.getSessionsModel());
            }
        });
        sessionDataViewModel.getSessionFeedBackResponse().observe(this,feedBackState -> {
            if (feedBackState.getFeedback() != null){
                handleFeedbackResponse(feedBackState.getFeedback());
            }
        });

    }

    private void handleFeedbackResponse(String feedback) {
        loginProgressBarFeedBack.setVisibility(View.GONE);

        txtSessionUserFeedback.setText("");

        Toast.makeText(getApplicationContext(),"Thank you for your feedback",
                Toast.LENGTH_SHORT).show();
    }

    private void getSessionData(String dayNumber, int sessionId) {
        sessionDataViewModel.getSessionDetails(dayNumber, sessionId);
    }

    private void handleFetchSessionData(SessionsModel sessionsModel) {
        if (sessionsModel != null) {
            sessionsModel1 = sessionsModel;
            //set the data on the view
            txtSessionFeedbackTitle.setText(sessionsModel.getTitle());

        }
    }

    private void handleDatabaseError(String databaseError) {
        Toast.makeText(getApplicationContext(), databaseError, Toast.LENGTH_SHORT).show();
    }

    @OnClick(R.id.fab)
    public void onViewClicked() {

        //get data from user and post them
        if(isFeedbackValid()){
            postUserFeedback(userFeedback);
        }

    }

    public boolean isFeedbackValid(){

        sessionFeedback = txtSessionUserFeedback.getText().toString().trim();
        boolean isValid;

        if (sessionFeedback.isEmpty()){
            txtSessionUserFeedback.setError("Session feedback cannot be empty");
            isValid = false;
        } else {

            isValid = true;
            txtSessionUserFeedback.setError(null);

            userFeedback = new SessionsUserFeedback();
            userFeedback.setSession_id(String.valueOf(sessionId));
            userFeedback.setDay_number(dayNumber);
            userFeedback.setSession_title(sessionsModel1.getTitle());
            userFeedback.setSession_feedback(sessionFeedback);
        }
        return isValid;
    }

    private void postUserFeedback(SessionsUserFeedback userFeedback) {

        loginProgressBarFeedBack.setVisibility(View.VISIBLE);
        sessionDataViewModel.sendSessionFeedBack(userFeedback);

    }
}

File: app/src/main/java/droiddevelopers254/droidconke/views/activities/SessionFeedBackActivity.kt
package droiddevelopers254.droidconke.views.activities

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import butterknife.ButterKnife
import droiddevelopers254.droidconke.R
import droiddevelopers254.droidconke.models.SessionsModel
import droiddevelopers254.droidconke.models.SessionsUserFeedback
import droiddevelopers254.droidconke.viewmodels.SessionDataViewModel
import kotlinx.android.synthetic.main.activity_session_feed_back.*
import kotlinx.android.synthetic.main.content_session_feed_back.*
import org.jetbrains.anko.toast
import org.koin.android.ext.android.inject


class SessionFeedBackActivity : AppCompatActivity() {
    var sessionId: Int = 0
    private var dayNumber :String = ""
    private lateinit var sessionsModel1 : SessionsModel
    private lateinit var userFeedback : SessionsUserFeedback
    private var sessionFeedback : String = ""

    private val sessionDataViewModel: SessionDataViewModel by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_session_feed_back)
        ButterKnife.bind(this)
        setSupportActionBar(toolbar)

       supportActionBar?.let {
           with(it) {
               setDisplayHomeAsUpEnabled(true)
               title = "Sessions Feedback"
           }
       }

        //get extras
        val extraIntent = intent
        sessionId = extraIntent.getIntExtra("sessionId", 0)
        dayNumber = extraIntent.getStringExtra("dayNumber")

        getSessionData(dayNumber, sessionId)

        //observe live data emitted by view model
        sessionDataViewModel.sessionData.observe(this, Observer{
            assert(it != null)
            when {
                it.databaseError != null -> handleDatabaseError(it.databaseError)
                else -> this.handleFetchSessionData(it.sessionsModel)
            }
        })
        sessionDataViewModel.getSessionFeedBackResponse().observe(this, Observer{
            handleFeedbackResponse(it.responseString)
        })

        fab.setOnClickListener {
            //get data from user and post them
            when {
                isFeedbackValid() -> postUserFeedback(userFeedback)
            }
        }

    }

    private fun handleFeedbackResponse(@Suppress("UNUSED_PARAMETER") feedback: String) {
        loginProgressBarFeedBack.visibility = View.GONE

        txtSessionUserFeedback.setText("")

        toast("Thank you for your feedback")
    }

    private fun getSessionData(dayNumber: String, sessionId: Int) {
        sessionDataViewModel.getSessionDetails(dayNumber, sessionId)
    }

    private fun handleFetchSessionData(sessionsModel: SessionsModel?) {
        if (sessionsModel != null) {
            sessionsModel1 = sessionsModel
            //set the data on the view
            txtSessionFeedbackTitle.text = sessionsModel.title

        }
    }

    private fun handleDatabaseError(databaseError: String) {
        Toast.makeText(applicationContext, databaseError, Toast.LENGTH_SHORT).show()
    }

    private fun isFeedbackValid(): Boolean {

        sessionFeedback = txtSessionUserFeedback.text.toString().trim()
        val isValid: Boolean

        when {
            sessionFeedback.isEmpty() -> {
                txtSessionUserFeedback.error = "Session feedback cannot be empty"
                isValid = false
            }
            else -> {

                isValid = true
                txtSessionUserFeedback.error = null

                userFeedback = SessionsUserFeedback(
                        user_id = "",
                        session_id = sessionId.toString(),
                        day_number = dayNumber,
                        session_title = sessionsModel1.title,
                        session_feedback = sessionFeedback
                )
            }
        }
        return isValid
    }

    private fun postUserFeedback(userFeedback: SessionsUserFeedback) {

        loginProgressBarFeedBack.visibility = View.VISIBLE
        sessionDataViewModel.sendSessionFeedBack(userFeedback)

    }
}